pkg/util/instanceselector/instanceselector.go

```diff
@@ -450,32 +471,33 @@ func instanceTypeSpecified(instanceType string) bool {
 	return instanceType != "" && !strings.ContainsRune(instanceType, '*')
 }
 
-func ResourcesToInstanceType(ps *api.PodSpec) (string, *bool, error) {
+func ResourcesToInstanceType(ps *api.PodSpec) (api.Architecture, string, *bool, error) {
 	if ps.Resources.ContainerInstance != nil && *ps.Resources.ContainerInstance {
-		return api.ContainerInstanceType, nil, nil
+		return ps.Architecture, api.ContainerInstanceType, nil, nil
 	}
 	if instanceTypeSpecified(ps.InstanceType) {
 		var sustainedCPU *bool
 		if ps.Resources.SustainedCPU != nil {
 			sustainedCPU = ps.Resources.SustainedCPU
 		}
-		return ps.InstanceType, sustainedCPU, nil
+		return ps.Architecture, ps.InstanceType, sustainedCPU, nil
 	}
 	if selector == nil {
 		msg := "fatal: instanceselector has not been initialized"
 		klog.Errorf(msg)
-		return "", nil, fmt.Errorf(msg)
+		return api.ArchUndefined, "", nil, fmt.Errorf(msg)
 	}
 	if ps.InstanceType == "" && noResourceSpecified(ps) {
-		return selector.defaultInstanceType, nil, nil
+		var arch = selector.architectureSelector(selector.defaultInstanceType)
+		return arch, selector.defaultInstanceType, nil, nil
 	}
 
-	instanceType, needsSustainedCPU := selector.getInstanceFromResources(ps.Resources, ps.InstanceType)
+	arch, instanceType, needsSustainedCPU := selector.getInstanceFromResources(ps.Resources, ps.InstanceType)
 	if instanceType == "" {
 		msg := "could not compute instance type from Spec.Resources. It's likely that the Pod.Spec.Resources specify an instance that doesnt exist in the cloud"
-		return "", nil, fmt.Errorf(msg)
+		return api.ArchUndefined, "", nil, fmt.Errorf(msg)
 	}
-	return instanceType, &needsSustainedCPU, nil
+	return arch, instanceType, &needsSustainedCPU, nil
```

Comment:
I'm not sure if we need to handle architecture so explicitly here.

Since an instance family/type also implicitly tells you the architecture (.e.g. `mac1` is always `arch_x86_64_mac`, `c6g` is always `arm`, `m5` is always `arch_x86_64`, etc) we can simply add a function to the `instanceselector` package that gives you back the architecture of the instance family or type. I think there's only one place where we need to look it up in the node controller; this way, we would not need to sprinkle and thread through architecture as another parameter/variable in the pod registry, the instance selector, the node controller etc.

Reply:
This is also required in the NodeScaler, and it doesn’t have access any reference to the instance selector as far as I know. That’s why I passed it through to the PodSpec.

Reply:
We can import the instanceselector package. I think it's better to keep the knowledge on architecture and how it relates to instance types confined to that package, instead of sprinkling it over a different packages.